crates/red_knot_python_semantic/src/types/call/arguments.rs

```diff
@@ -1,88 +1,179 @@
+use std::cell::Cell;
+use std::rc::Rc;
+
+use bitflags::bitflags;
+
 use super::Type;
 
-/// Typed arguments for a single call, in source order.
+/// Typed arguments for a single call, in source order, with an optional bound self/cls parameter.
 #[derive(Clone, Debug, Default)]
-pub(crate) struct CallArguments<'a, 'db>(Vec<Argument<'a, 'db>>);
+pub(crate) struct CallArguments<'a, 'db> {
+    bound_self: Option<Argument<'a, 'db>>,
+    arguments: Rc<[Argument<'a, 'db>]>,
+}
 
 impl<'a, 'db> CallArguments<'a, 'db> {
     /// Create a [`CallArguments`] with no arguments.
     pub(crate) fn none() -> Self {
-        Self(Vec::new())
+        Self {
+            bound_self: None,
+            arguments: vec![].into(),
+        }
     }
 
     /// Create a [`CallArguments`] from an iterator over non-variadic positional argument types.
     pub(crate) fn positional(positional_tys: impl IntoIterator<Item = Type<'db>>) -> Self {
         positional_tys
             .into_iter()
-            .map(Argument::Positional)
+            .map(|ty| Argument::positional().with_argument_type(ty))
             .collect()
     }
 
     /// Prepend an extra positional argument.
```

Comment:
Can we add a `## Panics` section to the documentation that explains that this method can only be called for `self` when `bound_self` is `None`. 

The alternative would be to make this method return an `Option` or a `Result` and use explicit `unwrap` calls. But that might be somewhat annoying if there are too many call sites. 

After reviewing the entire PR: It seems that this variant bubbles all the way up to `Type::try_*` methods. I think we should document this invariant on all methods where it is necessary or enforce it by changing the signature.

Reply:
This no longer panics with the new representation